source/fuzz/fuzzer_pass_permute_function_variables.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_permute_function_variables.h"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+#include "source/fuzz/fuzzer_context.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/fuzz/transformation_swap_function_variables.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassPermuteFunctionVariables::FuzzerPassPermuteFunctionVariables(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}  // Here we call parent constructor
+
+void FuzzerPassPermuteFunctionVariables::Apply() {
+  // Permuting OpVariable instructions in each function
+  for (auto& function : *GetIRContext()->module()) {
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfPermuteFunctionVariables())) {
+      continue;
+    }
+
+    auto first_block = function.entry().get();
+
+    std::vector<opt::Instruction*> variables;
+    for (auto& instruction : *first_block) {
+      if (instruction.opcode() == SpvOpVariable) {
+        variables.push_back(&instruction);
+      }
+    }
+
+    while (variables.size() >= 2) {
+      auto instruction_1 = GetFuzzerContext()->RemoveAtRandomIndex(&variables);
+      auto instruction_2 = GetFuzzerContext()->RemoveAtRandomIndex(&variables);
```

Comment:
This isn't quite right.

For example, the permutation (1 2 3) could never be realised this way, as it is the product of (1 2)(2 3), and your code would remove 1 and 2 if it picked the transposition (1 2), so that it would be impossible for (2 3) to be picked.

Instead I suggest you add another function to FuzzerContext called `GetChanceOfSwappingAnotherPairOfFunctionVariables` (with an associated member variable, initialized in the range 30-90%, say).

Then you could do:

```
do {
  uint32_t instruction_1_index = GetFuzzerContext()->RandomIndex(variables);
  uint32_t instruction_2_index = GetFuzzerContext()->RandomIndex(variables);
  ApplyTransformation(TransformationSwapFunctionVariables(variables[instrution_1_index]->result_id(), variables[instrution_2_index]->result_id()));
} while (GetFuzzerContext()->ChoosePercentage(GetFuzzerContext()->GetChanceOfSwappingAnotherPairOfFunctionVariables());
```

(The above is probably not quite right but hopefully gives you the idea.)

Reply:
Thank you Dr. @afd  for the suggestion, I have a question:  `ChoosePercentage` as I see 
```cpp
  bool FuzzerContext::ChoosePercentage(uint32_t percentage_chance) {
    assert(percentage_chance <= 100);
    return random_generator_->RandomPercentage() < percentage_chance;
  }
```
here I've probability to get 'true' for a long time or terminate from the first time, that right? if right, is that safe?

Reply:
If you are asking whether it's possible to get stuck in an infinite loop, then yes. In theory, it is possible. However, if the probability of doing another loop iteration is `k`, then the probability of doing `n` iterations is `k^(n-1)` (i.e. as the `n` grows, the probability of doing more iterations decreases since, per @afd's suggestion, `0.3 <= k <= 0.9`). Thus, in practice, it'll stop at some point.

It might be a good idea, though, to put an upper limit on the number of iterations we can do here.

Reply:
Also, you'll have to check that `instruction_1_index != instruction_2_index` since otherwise an assertion will fail in the `IsApplicable` method.

Reply:
Thanks for the clarification. :)

Reply:
I've followed this approach do you think it's good? @afd, @Vasniktel.
```c++
    uint32_t iterations_limit = 0;
        do {
          // Random Index template method which takes HasSize Method(std::vector) by
          // reference and returns the random index for instruction.
          uint32_t instruction_1_index = GetFuzzerContext()->RandomIndex(variables);
          uint32_t instruction_2_index = GetFuzzerContext()->RandomIndex(variables);
    
          if (instruction_1_index != instruction_2_index) {
            ApplyTransformation(TransformationSwapFunctionVariables(
                variables[instruction_1_index]->result_id(),
                variables[instruction_2_index]->result_id()));
          }
    
        } while (GetFuzzerContext()->ChoosePercentage(
                     GetFuzzerContext()
                         ->GetChanceOfSwappingAnotherPairOfFunctionVariables()) &&
                 iterations_limit++ <
                     25);  // I don't know 25 is a good limit or no!.
```
I don't know 25 is a good limit or no!!